build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -60,46 +61,39 @@ public void apply(Project project) {
-                    mainVersions.add(Integer.parseInt(sourcesetMatcher.group(1)));
-                }
-            }
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-
-        Collections.sort(mainVersions);
-        List<String> parentSourceSets = new ArrayList<>();
-        parentSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
+        List<Integer> mainVersions = findSourceVersions(project, MRJAR_MAIN_SOURCESET_PATTERN);
+        List<String> mainSourceSets = new ArrayList<>();
+        mainSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
         for (int javaVersion : mainVersions) {
-            String sourcesetName = "main" + javaVersion;
-            addMrjarSourceset(project, javaExtension, sourcesetName, parentSourceSets, javaVersion);
-            parentSourceSets.add(sourcesetName);
+            String sourcesetName = SourceSet.MAIN_SOURCE_SET_NAME + javaVersion;
+            addSourceset(project, javaExtension, sourcesetName, mainSourceSets, javaVersion);
+            mainSourceSets.add(sourcesetName);
         }
-    }
 
-    private void addMrjarSourceset(
-        Project project,
-        JavaPluginExtension javaExtension,
-        String sourcesetName,
-        List<String> parentSourceSets,
-        int javaVersion
-    ) {
-        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        for (String parentSourceSetName : parentSourceSets) {
-            GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
+        List<Integer> testVersions = findSourceVersions(project, MRJAR_TEST_SOURCESET_PATTERN);
+        List<String> testSourceSets = new ArrayList<>();
+        testSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
+        testSourceSets.add(SourceSet.TEST_SOURCE_SET_NAME);
+        for (int javaVersion : testVersions) {
+            String sourcesetName = SourceSet.TEST_SOURCE_SET_NAME + javaVersion;
+            if (mainVersions.contains(javaVersion)) {
+                // tests extend both their parent test source sets, and all of main sourcesets up to that version
+                testSourceSets.add(SourceSet.TEST_SOURCE_SET_NAME + javaVersion);
+            }
+            addSourceset(project, javaExtension, sourcesetName, testSourceSets, javaVersion);
+            testSourceSets.add(sourcesetName);
         }
 
+        configureMrjar(project, javaExtension, mainVersions, "main");
+    }
+
+    private void configureMrjar(Project project, JavaPluginExtension javaExtension, List<Integer> versions, String baseSourcesetName) {
         var jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME);
         jarTask.configure(task -> {
-            task.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+            for (var version : versions) {
+                SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(baseSourcesetName + version);
```

Comment:
Why are we attempting to create a source set here? Shouldn't this already exist? We also should definitely not be registering source sets in a task configuration action.

Reply:
Oops, copy/paste issue, I'll fix